### DISCUSSION TITLE
My vision of Aspire

### DISCUSSION THREAD

**Discussion by g-martin772** [microsoft/aspire#3967] [Ideas]
I don't know if there is an official roadmap for Aspire or something like that.
So I thought I might share my point of view on features that I would love to see:
- Hot Reloading -> What I do now to make this work is adding all projects as executable resources that call `dotnet watch` because restarting the entire stack every time you make a change is extremely slow. I suspect that the AddProject API just calls `dotnet run` in one way or another under the hood (but I haven't looked into it) so I thought the option to add a flag for whether you want to use run/watch might be a good idea.
- I know there are also other issues open regarding support for Restarting apps from the dashboard, but I still wanted to add it here. Being able to start/stop/restart specific services directly from the dashboard seems like the way to go for me
- Support for entity framework is there, but I think expanding on this might be a good idea since it is one of .NET's core features. Especially adding support for "hot swapping" databases. Basically being able to change the dataset/database or any container resource in general while the app continues to run. I know this might not be so easy to implement but it would certainly be cool.
- Another thing I wanted to mention is the way of dealing with credentials for containers like the SQL server password.  I don't know if there is a "correct" way to do this yet. Like being able to easily use local secret storage and also make it work well with deployment and CI/CD like Github Actions.
- An integrated way to manage and apply migrations. The current version of having a separate "DbManager" Project that applies the migrations works fine but setting this up for every db feels a bit tedious. 
- I know support for various Azure things is still work in progress and I hope this turns out great. Making any newbie like me able to deploy their distributed applications to Azure with basically one click would be amazing. 
- Managing container volumes from the Dashboard
- Maybe adding a way to visually represent how all the services are connected and all the pieces Aspire creates automatically to connect them and make everything work
- Maybe grouping resources together and having different "types" of run configurations like a "slim" version for local UI development that just connects a Blazor app to some dummy data and a "full" version that runs the real database and API and caching and all the other stuff. Since these distributed apps can get quite big quite quickly this would be amazing for local development because not everyone has a beefy workstation at hand (for example when working on a laptop)
- Local SMTP Mail Server Component

Other than that I think this is an amazing project! I have only been using it for a few weeks now so I am far from an expert. I hope my feedback is well placed in this Discussion forum. Any thoughts on that?

**Comment by davidfowl** [comment]
> Hot Reloading -> What I do now to make this work is adding all projects as executable resources that call dotnet watch because restarting the entire stack every time you make a change is extremely slow. I suspect that the AddProject API just calls dotnet run in one way or another under the hood (but I haven't looked into it) so I thought the option to add a flag for whether you want to use run/watch might be a good idea.

Still, lots of work to do here but normal hot reload works when you use F5. If you are using the command line then you can use dotnet watch on the apphost project. The 3rd pattern that we support from visual studio is using CTRL + F5 which enables launching without debugging while being able to change and rebuild individual projects without tearing everything down. This has been a constant source of feedback and we're still iterating on what the best incremental flow might be. We're also looking at future patterns where the IDE can keep around more state so that we can tear down less things (like the dashboard for e.g.).

> I know there are also other issues open regarding support for Restarting apps from the dashboard, but I still wanted to add it here. Being able to start/stop/restart specific services directly from the dashboard seems like the way to go for me

Definitely on the roadmap.

> Support for entity framework is there, but I think expanding on this might be a good idea since it is one of .NET's core features. Especially adding support for "hot swapping" databases. Basically being able to change the dataset/database or any container resource in general while the app continues to run. I know this might not be so easy to implement but it would certainly be cool.

cc @roji 

> Another thing I wanted to mention is the way of dealing with credentials for containers like the SQL server password. I don't know if there is a "correct" way to do this yet. Like being able to easily use local secret storage and also make it work well with deployment and CI/CD like Github Actions.

Depends on what you mean by this. If you're going to azure and using azd, then we have an out of the box solution for you. Otherwise, you fallback to providing values as part of your environment like you normally do in these scenarios.

> An integrated way to manage and apply migrations. The current version of having a separate "DbManager" Project that applies the migrations works fine but setting this up for every db feels a bit tedious.

Agree, for local development the overhead of building a standalone app is high. Though this method also works for deployment scenarios it feels a bit much when getting started.

> Managing container volumes from the Dashboard

Yes these should be visible for sure. Filed https://github.com/dotnet/aspire/issues/3969

> Maybe adding a way to visually represent how all the services are connected and all the pieces Aspire creates automatically to connect them and make everything work

On the roadmap.

> Maybe grouping resources together and having different "types" of run configurations like a "slim" version for local UI development that just connects a Blazor app to some dummy data and a "full" version that runs the real database and API and caching and all the other stuff. Since these distributed apps can get quite big quite quickly this would be amazing for local development because not everyone has a beefy workstation at hand (for example when working on a laptop)

Also on the roadmap. 

> Local SMTP Mail Server Component

You don't have to wait for us to build resources and components but this is a good suggestion! @mitchdenny is working on  documentation that shows how to build this https://github.com/dotnet/docs-aspire/pull/755

Keep the feedback coming!

**Reply by g-martin772** [reply]
> You don't have to wait for us to build resources and components but this is a good suggestion! @mitchdenny is working on documentation that shows how to build this https://github.com/dotnet/docs-aspire/pull/755

Thanks for the tipp! I have already built a few wrappers for the containers I use and it works great.

**Comment by g-martin772** [comment]
Regarding the database integration stuff with ef, I have a few more ideas.

Maybe we could have something like the DbManager Project built directly into the Dashboard. Where for each db you add in the AppHost you can optionally specify a Migrations Project to use for the db that then shows up in the dashboard. It would also be cool to have a SQL console where you can run SQL queries directly from the dashboard. For local development, you also usually have some seed data. Would be cool if we could set/reset the data with just a single click and confirm. I think the example also shows a `CreateStablePassword` method to create a password for the SQL container, but when using a Data Volume, it expects the same password when launching it for the second time so it just crashes if a new random pw is provided (I am not sure if this is what's supposed to happen but it does). Maybe having a built-in system that watches the volume state and just stores the pw somewhere as long as the volume exists, and when creating a new one also creates a new pw or something like that to improve the experience and have one thing less to think about.

**Reply by davidfowl** [reply]
> Maybe we could have something like the DbManager Project built directly into the Dashboard.

Not sure if we should build it into the aspire dashboard. One of the ideas was to ship a migration UI as part of EF that could be used in aspire (and anywhere else). This would follow a similar pattern to the existing database tools we wire up today (phpmyadmin, mongoexpress, pgadmin etc). 

>  I think the example also shows a CreateStablePassword method to create a password for the SQL container, but when using a Data Volume, it expects the same password when launching it for the second time so it just crashes if a new random pw is provided (I am not sure if this is what's supposed to happen but it does). Maybe having a built-in system that watches the volume state and just stores the pw somewhere as long as the volume exists, and when creating a new one also creates a new pw or something like that to improve the experience and have one thing less to think about.

I don't think it needs to be that complex. We plan to do this by default for the containers that rely on it when a volume is used https://github.com/dotnet/aspire/issues/1151

**Reply by roji** [reply]
Yeah, I don't think Aspire should contain a full-fledged SQL tool, including a console allowing running arbitrary queries etc. - that really seems out-of-scope and would be a huge effort. There already are very good database-specific database management tools out there - such as pgadmin - we should not be attempting to reimplement such stuff.

> Support for entity framework is there, but I think expanding on this might be a good idea since it is one of .NET's core features. Especially adding support for "hot swapping" databases. Basically being able to change the dataset/database or any container resource in general while the app continues to run. I know this might not be so easy to implement but it would certainly be cool.

Can you elaborate on exactly what this means and why you think it's needed, in what scenarios etc.?

**Reply by g-martin772** [reply]
Lets say I have a few different datasets for testing, being able to have multiple volumes/data sets and the just selecting the one you want whould be cool in my opinion.

I am also not only thinking about sql but also other container resources with volumes, because in development I end up deleting/resetting them quite often. For example when data becomes invalid or you just want to reset the config.

Are there any plans on adding interactive terminals to the dashbord so one can interact with containers right from the dashboard? It doesnt have to be a idk crazy facy terminal. Or maybe we could just display connection strings for sql and containers so you can just copy paste them into your terminal of choice.

Thanks for the feedback so far!

**Reply by g-martin772** [reply]
> Yeah, I don't think Aspire should contain a full-fledged SQL tool, including a console allowing running arbitrary queries etc. - that really seems out-of-scope and would be a huge effort. There already are very good database-specific database management tools out there - such as pgadmin - we should not be attempting to reimplement such stuff.

Yeah, I think you are right.

**Reply by roji** [reply]
> Lets say I have a few different datasets for testing, being able to have multiple volumes/data sets and the just selecting the one you want whould be cool in my opinion.

So it sounds like you want the ability to manually create database and seed them, choosing seeding "sets" from the UI. That's once again quite a feature to design and implement (in a way that satisifies everyone etc.).

**Reply by g-martin772** [reply]
If thats not a target for Aspire thats fine with me. When implementing a UI to show the container volumes, as mentioned above, at least a little "reset" button for data volumes whould be nice if that is possible.

**Reply by roji** [reply]
@g-martin772 that again has complex interactions with the specific application being managed. For example, different applications manage their database data/seeding differently using whatever code mechanism they choose, and this would have to somehow hook into that.

**Reply by g-martin772** [reply]
I am not specifically talking about databases. I mean for all container volumes in general. Since Aspire is abstracting all of the Docker interaction for managing containers, networks and volumes I thought why not that too?

**Reply by roji** [reply]
It's easy enough to kill and recreate a database container, but what about the data in it? The application (another container) is presumably already running and has seeded/manipulated the database, and it would now be replaced with an empty one (no tables whatsoever). So in most cases this probably won't work as-is, and would require at least restarting the application as well (which may be fine).

**Reply by g-martin772** [reply]
Since the ability to restart resources from the dashboard is planned to be added, a system that detects what services are connected in what way, so that dependent services get restarted too, will probably be implemented anyway.

**Reply by roji** [reply]
Maybe; if that's the case, there doesn't seem to be anything specifically related to database here though - it would be a purely container chained restart feature, nothing more.

**Reply by davidfowl** [reply]
@roji I think an interactive database migration UI for EF would be a useful interactive tool we could make part of the core aspire experience. It would be the equivalent of using the command line ef tool.

**Reply by roji** [reply]
Sure, we could definitely think about something like that... We'd ideally make it an independent UI component that would then be integratable in the Aspire dashboard?

**Reply by davidfowl** [reply]
Yep that’s the thought.
